error: Unsupported architecture [macOS->Android]

This issue related to macOS.

While I trying to build my script for Android using `cargo build --target aarch64-linux-android` I getting this (architecture not supported) compiler error:

```
error: failed to run custom build command for `gdnative-sys v0.8.0`

Caused by:
  process didn't exit successfully: `/Users/arsenius/Projects/godot/FC/scripts/target/debug/build/gdnative-sys-ef8bc7869be9b0c8/build-script-build` (exit code: 101)
--- stdout
cargo:warning=couldn't execute `llvm-config --prefix` (error: No such file or directory (os error 2))
cargo:warning=set the LLVM_CONFIG_PATH environment variable to a valid `llvm-config` executable

--- stderr
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/cdefs.h:807:2: error: Unsupported architecture
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/machine/_types.h:34:2: error: architecture not supported
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types.h:55:9: error: unknown type name '__int64_t'
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types.h:56:9: error: unknown type name '__int32_t'
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types.h:57:9: error: unknown type name '__int32_t'
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types.h:60:9: error: unknown type name '__uint32_t'
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types.h:61:9: error: unknown type name '__uint32_t'
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types.h:62:9: error: unknown type name '__uint64_t'
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types.h:68:9: error: unknown type name '__darwin_natural_t'
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types.h:70:9: error: unknown type name '__uint16_t'
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types.h:71:9: error: unknown type name '__int64_t'
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types.h:72:9: error: unknown type name '__int32_t'
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types.h:73:9: error: unknown type name '__uint32_t'
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types.h:74:9: error: unknown type name '__int32_t'
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types.h:75:9: error: unknown type name '__uint32_t'
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types.h:76:9: error: unknown type name '__uint32_t'
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/machine/types.h:37:2: error: architecture not supported
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types/_intptr_t.h:32:9: error: unknown type name '__darwin_intptr_t'
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/_types.h:43:9: error: unknown type name '__uint32_t'
fatal error: too many errors emitted, stopping now [-ferror-limit=]
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/cdefs.h:807:2: error: Unsupported architecture, err: true
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/machine/_types.h:34:2: error: architecture not supported, err: true
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types.h:55:9: error: unknown type name '__int64_t', err: true
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types.h:56:9: error: unknown type name '__int32_t', err: true
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types.h:57:9: error: unknown type name '__int32_t', err: true
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types.h:60:9: error: unknown type name '__uint32_t', err: true
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types.h:61:9: error: unknown type name '__uint32_t', err: true
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types.h:62:9: error: unknown type name '__uint64_t', err: true
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types.h:68:9: error: unknown type name '__darwin_natural_t', err: true
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types.h:70:9: error: unknown type name '__uint16_t', err: true
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types.h:71:9: error: unknown type name '__int64_t', err: true
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types.h:72:9: error: unknown type name '__int32_t', err: true
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types.h:73:9: error: unknown type name '__uint32_t', err: true
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types.h:74:9: error: unknown type name '__int32_t', err: true
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types.h:75:9: error: unknown type name '__uint32_t', err: true
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types.h:76:9: error: unknown type name '__uint32_t', err: true
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/machine/types.h:37:2: error: architecture not supported, err: true
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types/_intptr_t.h:32:9: error: unknown type name '__darwin_intptr_t', err: true
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/_types.h:43:9: error: unknown type name '__uint32_t', err: true
fatal error: too many errors emitted, stopping now [-ferror-limit=], err: true
thread 'main' panicked at 'Unable to generate bindings: ()', /Users/arsenius/.cargo/registry/src/github.com-1ecc6299db9ec823/gdnative-sys-0.8.0/build.rs:110:24
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

warning: build failed, waiting for other jobs to finish...
error: build failed
```

I used `make_standalone_toolchain.py` script to generate standalone NDKs for arm64, arm and x86 and then created `~/.cargo/config` file with following data:

```
[target.aarch64-linux-android]
ar = "<path_to_standalone_ndk>/arm64/bin/aarch64-linux-android-ar"
linker = "<path_to_standalone_ndk>/arm64/bin/aarch64-linux-android-clang"

[target.armv7-linux-androideabi]
ar = "<path_to_standalone_ndk>/arm/bin/arm-linux-androideabi-ar"
linker = <path_to_standalone_ndk>/arm/bin/arm-linux-androideabi-clang"

[target.i686-linux-android]
ar = "<path_to_standalone_ndk>/x86/bin/i686-linux-android-ar"
linker = "<path_to_standalone_ndk>/x86/bin/i686-linux-android-clang"
```

How to solve this issue?

> Which version of godot-rust are you using? It seems like the MacOS platform headers are being erroneously included due to the platform you're building from. This is a known problem that _should_ have been fixed in #286, which is included in 0.8.0. If you're not using it, could you try and see if the problem is present on 0.8.0?
> 
> Also, there is a chapter about building for Android written by @tommywalkie in the WiP user guide: https://github.com/GodotNativeTools/godot-rust/blob/book/book/src/exporting/android.md. Could you try and see if following the instructions there changes anything?

I'm using this version:
![godot-rust version](https://i.ibb.co/VgcMcPb/Screenshot-2020-05-15-at-3-45-21-PM.png)

Yes the chapter is really helpful and it helps to resolve issue with `godot_android.h:37:10: fatal error: 'jni.h' file not found` when I target `i686-linux-android` but with target `aarch64-linux-android` 

Maybe because llvm not installed on my Mac?

> > It took ages to install llvm on my mac
> 
> @atonamy It should take ~2.6GB so it makes sense.
> 
> > version 10 is ok?
> 
> Apparently [yes](https://github.com/tommywalkie/sample-godot-rust-app/runs/677600980?check_suite_focus=true#step:9:4)

The link you just provided please check the log failing build for `arm7`, `arm64` [here](https://github.com/tommywalkie/sample-godot-rust-app/runs/677600980?check_suite_focus=true#step:12:22)
is actually exactly the same problem what I'm still facing. 
Is only helped to resolve warning but didn't fix the main error:
`/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/cdefs.h:807:2: error: Unsupported architecture`

> @atonamy Please keep the discussion thread on topic. If you have encountered a different problem, please create another issue for it.
> 
> As for the original issue, I've added a fix in #345. Please try and see if it works on your machine.

I just tried, at this time compiler asked me JAVA_HOME and ANDROID_SDK_ROOT so I provided those variables, but  same problem reamain:
```
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/cdefs.h:807:2: error: Unsupported architecture
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/machine/_types.h:34:2: error: architecture not supported
```

Ok I will create separate issue for not working Android x86_64 gdnative binding  from Mac OS

--- a/gdnative-sys/build.rs
+++ b/gdnative-sys/build.rs
@@ -11,7 +11,7 @@ fn main() {
 }
 
 mod header_binding {
-    use std::path::PathBuf;
+    use std::path::{Path, PathBuf};
 
     fn osx_include_path() -> Result<String, std::io::Error> {
         use std::process::Command;
@@ -43,6 +43,113 @@ mod header_binding {
         let directory = format!("{}/{}/{}", prefix, infix, suffix);
 
         Ok(directory)
+    }
+
+    fn add_android_include_paths(mut builder: bindgen::Builder) -> bindgen::Builder {
+        let target_os = std::env::var("CARGO_CFG_TARGET_OS").unwrap();
+        let target_triple = std::env::var("TARGET").unwrap();
+
+        assert_eq!("android", &target_os);
+
+        let java_home =
+            std::env::var("JAVA_HOME").expect("JAVA_HOME and ANDROID_SDK_ROOT must be set");
+        let java_home = Path::new(&java_home).to_path_buf();
+        let android_sdk_root =
+            std::env::var("ANDROID_SDK_ROOT").expect("JAVA_HOME and ANDROID_SDK_ROOT must be set");
+        let android_sdk_root = Path::new(&android_sdk_root).to_path_buf();
+
+        // Note: cfg!(target_os) and cfg!(target_arch) refer to the target of the build script:
+        // in other words, the host machine instead of the target of gdnative-sys. They are confusing
+        // and have been erroneously used for target platforms in this library in the past. Make sure
+        // to double-check them wherever they occur.
+
+        if !cfg!(target_arch = "x86_64") {
+            panic!("unsupported host architecture: build from x86_64 instead");
+        }
+
+        builder = builder
+            .clang_arg("-I")
+            .clang_arg(Path::join(&java_home, "include/").to_string_lossy());
+
+        builder = builder.clang_arg("-I").clang_arg(
+            Path::join(
+                &java_home,
+                format!("include/{}/", {
+                    if cfg!(target_os = "windows") {
+                        "win32"
+                    } else if cfg!(target_os = "macos") {
+                        "darwin"
+                    } else if cfg!(target_os = "linux") {
+                        "linux"
+                    } else {
+                        panic!("unsupported host OS: build from Windows, MacOS, or Linux instead");
+                    }
+                }),
+            )
+            .to_string_lossy(),
+        );
+
+        builder = builder.clang_arg("-I").clang_arg(
+            Path::join(&android_sdk_root, "ndk-bundle/sysroot/usr/include").to_string_lossy(),
+        );
+        builder = builder.clang_arg("-I").clang_arg(
+            Path::join(
+                &android_sdk_root,
+                "ndk-bundle/sources/cxx-stl/llvm-libc++/include",
+            )
+            .to_string_lossy(),
+        );
+        builder = builder.clang_arg("-I").clang_arg(
+            Path::join(
+                &android_sdk_root,
+                "ndk-bundle/sources/cxx-stl/llvm-libc++abi/include",
+            )
+            .to_string_lossy(),
+        );
+        builder = builder.clang_arg("-I").clang_arg(
+            Path::join(
+                &android_sdk_root,
+                "ndk-bundle/sources/android/support/include",
+            )
+            .to_string_lossy(),
+        );
+
+        let host_tag = {
+            if cfg!(target_os = "windows") {
+                "windows-x86_64"
+            } else if cfg!(target_os = "macos") {
+                "darwin-x86_64"
+            } else if cfg!(target_os = "linux") {
+                "linux-x86_64"
+            } else {
+                panic!("unsupported host OS: build from Windows, MacOS, or Linux instead");
+            }
+        };
+
+        builder = builder.clang_arg("-I").clang_arg(
+            Path::join(
+                &android_sdk_root,
+                format!(
+                    "ndk-bundle/toolchains/llvm/prebuilt/{}/sysroot/usr/include",
+                    &host_tag
+                ),
+            )
+            .to_string_lossy(),
+        );
+
+        builder = builder.clang_arg("-I").clang_arg(
+            Path::join(
+                &android_sdk_root,
+                format!(
+            "ndk-bundle/toolchains/llvm/prebuilt/{host}/sysroot/usr/include/{target_triple}",
+            host = &host_tag,
+            target_triple = &target_triple,
+        ),
+            )
+            .to_string_lossy(),
+        );
+
+        builder
     }
 
     fn is_travis_ci() -> bool {
@@ -105,6 +212,10 @@ mod header_binding {
                 .clang_arg("-fms-extensions")
                 .clang_arg("-fmsc-version=1300")
                 .clang_arg("-D_M_X64=100");
+        }
+
+        if target_os == "android" {
+            builder = add_android_include_paths(builder);
         }
 
         let bindings = builder.generate().expect("Unable to generate bindings");

--- a/gdnative-sys/src/lib.rs
+++ b/gdnative-sys/src/lib.rs
@@ -1,6 +1,7 @@
 #![allow(
     non_upper_case_globals,
     non_camel_case_types,
+    non_snake_case,
     improper_ctypes,
     clippy::style
 )]

